rustls/src/quic.rs

```diff
@@ -108,7 +145,62 @@ impl PacketKey {
                 &[],
             )),
             iv: hkdf_expand(secret, IvLen, b"quic iv", &[]),
+            suite,
+        }
+    }
+
+    /// Encrypt a QUIC packet
+    ///
+    /// `buf` contains the header, payload and space for the AEAD tag, with the payload starting at
+    /// `header_len` and ending at `tag_len()` before the end of `buf`.
+    pub fn encrypt(&self, packet: u64, buf: &mut [u8], header_len: usize) -> Result<(), Error> {
+        let (header, payload) = buf.split_at_mut(header_len);
+        let (payload, tag_storage) =
+            payload.split_at_mut(payload.len() - self.key.algorithm().tag_len());
+        let aad = aead::Aad::from(header);
+        let nonce = self.iv.nonce_for(packet);
+        let tag = self
+            .key
+            .seal_in_place_separate_tag(nonce, aad, payload)
+            .map_err(|_| Error::EncryptError)?;
+        tag_storage.copy_from_slice(tag.as_ref());
+        Ok(())
+    }
+
+    /// Decrypt a QUIC packet
+    ///
+    /// Returns the length of the decrypted payload.
+    pub fn decrypt(&self, packet: u64, header: &[u8], payload: &mut [u8]) -> Result<usize, Error> {
+        if payload.len() < self.key.algorithm().tag_len() {
+            return Err(Error::DecryptError);
         }
+
+        let payload_len = payload.len();
+        let aad = aead::Aad::from(header);
+        let nonce = self.iv.nonce_for(packet);
+        self.key
+            .open_in_place(nonce, aad, payload)
+            .map_err(|_| Error::DecryptError)?;
+        Ok(payload_len - self.key.algorithm().tag_len())
+    }
+
+    /// Number of times the packet key can be used without sacrificing confidentiality
+    ///
+    /// See https://www.rfc-editor.org/rfc/rfc9001.html#name-confidentiality-limit.
+    pub fn confidentiality_limit(&self) -> u64 {
```

Comment:
1. Why don't `encrypt()` and `decrypt()` enforce these limits?
2. If `encrypt()` and `decrypt()` were to enforce these limits then would we still need to expose them?
3. Should we be enforcing these limits within Rustls in TLS?

These functions are called frequently? Probably should be `#[inline]`.

Reply:
The QUIC implementation initiates a key update when the confidentiality limit is approaching, but not yet reached, so checking the limit only during `encrypt` would not be sufficient. Either it needs to be exposed, or rustls would have to make a call on what "approaching" means (which might be difficult to do in a way that satisfies every user) and that state would have to be polled.

The integrity limit could perhaps be checked during `decrypt`, but we'd need a separate error case for it so we can respond appropriately (by closing the connection) while still silently dropping packets that couldn't be authenticated otherwise.

Reply:
> 3. Should we be enforcing these limits within Rustls in TLS?

Assuming you mean TCP-based TLS here as opposed to QUIC, RFC 8446 doesn't seem to mention anything like the stuff layed out in RFC 9001. In any case, IMO it would be out of scope for this PR.

Reply:
OK. The same logic applies to both TLS and QUIC and we should factor this stuff out into a common place in the future, so that a TLS 1.3 application can also do a key update and so a TLS 1.2 application can stop if/when the keys are exhausted.

Reply:
I filed issue #755 to follow up on this. From my perspective, this is resolved.